internal/js_parser/js_parser.go

```diff
@@ -14351,3 +14373,780 @@ func (p *parser) toAST(parts []js_ast.Part, hashbang string, directive string) j
 		TopLevelAwaitKeyword: p.topLevelAwaitKeyword,
 	}
 }
+
+func addShopifyI18n(p *parser, parts *[]js_ast.Part, translationsDir string) {
+	options := p.options
+
+	// Set up `import _en from './path/to/translations/en.json`
+	importRecordIndex := p.addImportRecord(ast.ImportStmt, p.lexer.Loc(), translationsDir + "/en.json", nil)
+	defaultName := js_ast.LocRef{Loc: p.lexer.Loc(), Ref: p.newSymbol(js_ast.SymbolOther, "_en")}
+	p.moduleScope.Generated = append(p.moduleScope.Generated, defaultName.Ref)
+	p.recordUsage(defaultName.Ref)
+
+	(*parts)[0].Stmts = append(
+		[]js_ast.Stmt{{Data: &js_ast.SImport{
+			DefaultName: &defaultName,
+			ImportRecordIndex: importRecordIndex,
+		}}},
+		(*parts)[0].Stmts...,
+	)
+
+	// Set up `fetch` for remote translation requests
+	fetchRef, fetchRefExists := p.moduleScope.Members["fetch"]
+	if !fetchRefExists {
+		p.declareSymbol(js_ast.SymbolUnbound, p.lexer.Loc(), "fetch")
+		fetchRef, _ = p.moduleScope.Members["fetch"]
+	}
+
+	// Create an argument for the `fallback: (e) => {}` callback
+	localeArg := js_ast.LocRef{Loc: p.lexer.Loc(), Ref: p.newSymbol(js_ast.SymbolOther, "_translationsLocale")}
+	(*p.shopifyI18nScope).Generated = append(p.moduleScope.Generated, localeArg.Ref)
+
+	p.shopifyI18nCallRef.Args = append(
+		p.shopifyI18nCallRef.Args,
+		js_ast.Expr{
```

Comment:
How much am I going to regret all of this boilerplate? Soon! How much more am I going to regret not writing tests for it? 😞

Reply:
There's probably a way to do this by parsing little code blocks (similar to the way runtimeAST is generated). This works for now, and my rationale is that i18n + hmr code changes very rarely.